[Feature] Introduce predicate for sstable

srlch opened the pull request:
## What I'm doing:
This pr introduce predicate which can be evaluted for key for sstable of lake persistent index. This pr is prerequisite pr for tablet split. We only introduce the predicate definition and the read path filtering will be added in next pr

Fixes #59134

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.5
  - [ ] 3.4
  - [ ] 3.3

xiangguangyxg commented on be/src/storage/lake/sstable_predicate.h:
```diff
@@ -0,0 +1,77 @@
+    RowToChunkConverterUPtr _converter;
+};
+
+class RowToChunkConverter {
+public:
+    RowToChunkConverter(Schema pkey_schema, ChunkUniquePtr& pk_chunk, MutableColumnPtr& pk_column)
+            : _pkey_schema(pkey_schema), _pk_chunk(std::move(pk_chunk)), _pk_column(std::move(pk_column)) {}
+    ~RowToChunkConverter() = default;
+
+    static StatusOr<RowToChunkConverterUPtr> create(const TabletSchemaPB& tablet_schema_pb);
+    // Convert row data from sstable to a chunk
+    StatusOr<ChunkUniquePtr> convert_to_chunk(SstablePredicate::Type type, const std::string& row);
+
+private:
+    StatusOr<ChunkUniquePtr> _convert_key_to_chunk(const std::string& row);
+
+    // convert context for key
+    Schema _pkey_schema;
+    ChunkUniquePtr _pk_chunk;
+    MutableColumnPtr _pk_column;
+};
```
Put this class to a separared source code file

xiangguangyxg commented on be/src/storage/lake/sstable_predicate.h:
```diff
@@ -0,0 +1,77 @@
+class SstablePredicate {
+public:
+    // indicate predicate should be evaluated on key or value
+    enum class Type {
+        KEY = 0,
+        VALUE = 1,
+    };
```
not needed ?

srlch committed:
[Feature] Introduce predicate for sstable

## What I'm doing:
This pr introduce predicate which can be evaluted for key for sstable of lake persistent index.
This pr is prerequisite pr for tablet split. We only introduce the predicate definition and the read
path filtering will be added in next pr

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
fix

Signed-off-by: srlch <linzichao@starrocks.com>

wyb commented on be/src/storage/lake/sstable_predicate.h:
```diff
@@ -0,0 +1,46 @@
+    Status evaluate(const std::string& row, uint8_t* selection);
+
+private:
+    Status _init(const PersistentIndexSstablePredicatePB& predicate_pb, KeyToChunkConverterUPtr& converter);
```
```suggestion
    Status init(const PersistentIndexSstablePredicatePB& predicate_pb, KeyToChunkConverterUPtr& converter);
```
